docs/configuration.md
---
id: configuration
title: Configuration
sidebar_label: Configuration
---


Configuration of rest-hapi is handled through the ``RestHapi.config`` object. See the [introduction](introduction.md#using-the-plugin) for an example.  Below is a description of the current configuration options/properties.

## appTitle
```javascript
/**
 * Your app title goes here.
 * @type {string}
 */
config.appTitle = 'rest-hapi API'
```

## version
```javascript
/**
 * Your app version goes here.
 * @type {string}
 */
config.version = '1.0.0'
```

## absoluteModelPath
```javascript
/**
 * Flag signifying whether the absolute path to the models directory is provided
 * default: false
 * @type {boolean}
 */
config.absoluteModelPath = false
```

## modelPath
```javascript
/**
 * Path to the models directory
 * default: 'models'
 * @type {string}
 */
config.modelPath = 'models'
```

## absoluteApiPath
```javascript
/**
 * Flag signifying whether the absolute path to the api directory is provided
 * @type {boolean}
 */
config.absoluteApiPath = false
```

## apiPath
```javascript
/**
 * Path to the directory for additional endpoints
 * default: 'api'
 * @type {string}
 */
config.apiPath = 'api'
```

## cors
### additionalHeaders
### additionalExposedHeaders
```javascript
/**
 * Cors settings for generated endpoints. Can be set to false to disable.
 * @type {{additionalHeaders: string[], additionalExposedHeaders: string[]}}
 */
config.cors = {
  additionalHeaders: [],
  additionalExposedHeaders: []
}
```

## mongo
### URI
```javascript
/**
 * Mongo settings
 * - config.mongo.URI = 'mongodb://localhost/rest_hapi'; (local db, default)
 * - config.mongo.options = {} (mongoose options)
 */
config.mongo.URI = 'mongodb://localhost/rest_hapi'
config.mongo.options = {}
```

## authStrategy
```javascript
/**
 * Authentication strategy to be used for all generated endpoints.
 * Set to false for no authentication.
 * default: false
 * @type {boolean/string}
 */
config.authStrategy = false
```

## embedAssociations
```javascript
/**
 * If set to false, MANY_MANY associations (including linking model data) will be saved in their own collection in th db.  This is useful if a single document
 * will be associated with many other documents, which could cause the document size to become very large. For example,
 * a business might be associated with thousands of users.
 *
 * Embedding the associations will be more efficient for population/association queries but less efficient for memory/document size.
 *
 * This setting can be individually overwritten by setting the "embedAssociation" association property.
 * default: false
 * @type {boolean}
 */
config.embedAssociations = false
```

## enableCreatedAt
## enableUpdatedAt
## enableDeletedAt
## enableCreatedBy
## enableUpdatedBy
## enableDeletedBy
```javascript
/**
 * MetaData options:
 * - createdAt: (default: true) date specifying when the document was created.
 * - updatedAt: (default: true) date specifying when the document was last updated.
 * - deletedAt: (default: true) date specifying when the document was soft deleted.
 * - createdBy: (default: false) _id of user that created the document.
 * - updatedBy: (default: false) _id of user that last updated the document.
 * - deletedBy: (default: false) _id of user that soft deleted the document.
 */
config.enableCreatedAt = true
config.enableUpdatedAt = true
config.enableDeletedAt = true
config.enableCreatedBy = false
config.enableUpdatedBy = false
config.enableDeletedBy = false
```

## enableWhereQueries
```javascript
/**
 * Enables raw $where mongoose queries
 * !!WARNING!!: This feature is meant for development ONLY and NOT in production as it provides direct query access to your database.
 * default: false
 * @type {boolean}
 */
config.enableWhereQueries = false
```

## enableDuplicateFields
```javascript
/**
 * Enables fields from an associated model to be duplicated. Similar to permanently embedding an associated field within
 * the parent model schema. Useful if a parent model needs to be searchable or sortable by an association's field.
 * default: false
 * @type {boolean}
 */
config.enableDuplicateFields = false
```

## trackDuplicatedFields
```javascript
/**
 * When true, duplicated fields will update whenever the original field is updated.
 * WARNING: This feature can make updates very resource intensive if many documents are duplicating the original field.
 * default: false
 * @type {boolean}
 */
config.trackDuplicatedFields = false
```

## enableAuditLog
```javascript
/**
 * When enabled, all create, update, associate, and delete events are recorded in an auditLog collection.
 * default: true
 * @type {boolean}
 */
config.enableAuditLog = true
```

## auditLogScope
```javascript
/**
 * Values added here will be applied to the scope of the auditLog endpoint.
 * default: []
 * @type {Array}
 */
config.auditLogScope = []
```

## auditLogTTL
```javascript
/**
 * Specifies the TTL (time to live/lifetime/expiration) of auditLog documents. Accepts values in seconds unless specified
 * (Ex: 60 = 60 seconds, '1m' = 1 minute, or '1d' = 1 day)
 * See: http://nicoll.io/mongottl/
 * default: null (does not expire)
 * @type {string}
 */
config.auditLogTTL = null
```

## enablePolicies
```javascript
/**
 * Enables policies via mrhorse (https://github.com/mark-bradshaw/mrhorse).
 * default: false
 * @type {boolean}
 */
config.enablePolicies = false
```

## absolutePolicyPath
```javascript
/**
 * Flag signifying whether the absolute path to the policies directory is provided.
 * default: false
 * @type {boolean}
 */
config.absolutePolicyPath = false
```

## policyPath
```javascript
/**
 * Path to the directory for mrhorse policies (https://github.com/mark-bradshaw/mrhorse).
 * default: 'policies'
 * @type {string}
 */
config.policyPath = 'policies'
```

## enableDocumentScopes
```javascript
/**
 * Enables document level authorization.
 * default: true
 * @type {boolean}
 */
config.enableDocumentScopes = true
```

## authorizeDocumentCreator
```javascript
/**
 * If true, modifies the root scope of any document to allow access to the document's creator.
 * The scope value added is in the form: "user-{_id}" where "{_id}" is the _id of the user.
 * NOTE:
 * - This assumes that your authentication credentials (request.auth.credentials) will contain either
 * a "user" object with a "_id" property, or the user's _id stored in a property defined by "config.userIdKey".
 * - This also assumes that the user creating the document will have "user-{_id}" within their scope.
 * - Requires "config.enableDocumentScopes" to be "true".
 * - This setting can be individually overwritten by setting the "authorizeDocumentCreator" routeOptions property.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreator = false
```

## authorizeDocumentCreatorToRead
```javascript
/**
 * Same as "authorizeDocumentCreator", but modifies the "readScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToRead = false
```

## authorizeDocumentCreatorToUpdate
```javascript
/**
 * Same as "authorizeDocumentCreator", but modifies the "updateScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToUpdate = false
```

## authorizeDocumentCreatorToDelete
```javascript
/**
 * Same as "authorizeDocumentCreator", but modifies the "deleteScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToDelete = false
```

## authorizeDocumentCreatorToAssociate
```javascript
/**
 * Same as "authorizeDocumentCreator", but modifies the "associateScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToAssociate = false
```

## userIdKey
```javascript
/**
 * This is the path/key to the user _id stored in your request.auth.credentials object.
 * default: "user._id"
 * @type {string}
 */
config.userIdKey = 'user._id'
```

## enableDocumentScopeFail
```javascript
/**
 * Determines what action takes place when one or more document scope checks fail for requests dealing with multiple
 * documents (Ex: deleteMany or list). Options are:
 * - true: if one or more documents fail, the request responds with a 403.
 * - false: documents that don't pass are simply removed from the request (Ex: not deleted or not retrieved)
 * default: false
 * @type {boolean}
 */
config.enableDocumentScopeFail = false
```

## enableTextSearch
```javascript
/**
 * Flag specifying whether to text index all string fields for all models to enable text search.
 * WARNING: enabling this adds overhead to add inserts and updates, as well as added storage requirements.
 * default: false.
 * @type {boolean}
 */
config.enableTextSearch = false
```

## enableSoftDelete
## filterDeletedEmbeds
```javascript
/**
 * Soft delete options
 * - enableSoftDelete: adds "isDeleted" property to each model. Delete endpoints set "isDeleted" to true
 * unless the payload contains { hardDelete: true }, in which case the document is actually deleted (default false)
 * - filterDeletedEmbeds: if enabled, associations with "isDeleted" set to true will not populate (default false)
 * NOTE: this option is known to be buggy
 * @type {boolean}
 */
config.enableSoftDelete = false
config.filterDeletedEmbeds = false
```


## enableQueryValidation
## enablePayloadValidation
## enableResponseValidation
```javascript
/**
 * Validation options:
 * default: true
 * @type {boolean}
 */
config.enableQueryValidation = true
config.enablePayloadValidation = true
config.enableResponseValidation = true
```

## enableMongooseRunValidators
```javascript
/**
 * Mongoose validation options:
 * - enableMongooseRunValidators: enables the runValidators option in Mongoose update calls
 * <http://mongoosejs.com/docs/validation.html#update-validators>
 * default: false
 * @type {boolean}
 */
config.enableMongooseRunValidators = false
```

## enableResponseFail
```javascript
/**
 * Determines the hapi failAction of each response. Options are:
 * - true: responses that fail validation will return a 500 error.
 * - false: responses that fail validation will just log the offense and send the response as-is.
 * default: false
 * @type {boolean}
 */
config.enableResponseFail = false
```

## generateRouteScopes
```javascript
/**
 * If set to true, (and authStrategy is not false) then endpoints will be generated with pre-defined
 * scopes based on the model definition.
 * default: false
 * @type {boolean}
 */
config.generateRouteScopes = false
```

## logScopes
```javascript
/**
 * If set to true, the scope for each endpoint will be logged when then endpoint is generated.
 * default: false
 * @type {boolean}
 */
config.logScopes = false
```

## logRoutes
```javascript
/**
 * If set to true, each route will be logged as it is generated.
 * default: false
 * @type {boolean}
 */
config.logRoutes = false
```

## loglevel
```javascript
/**
 * Log level options:
 * - INTERNAL use it for logging calls and other internal stuff
 * - DEBUG recommended to use it for debugging applications
 * - NOTE development verbose information (default)
 * - INFO minor information
 * - LOG significant messages
 * - WARNING really important stuff
 * - ERROR application business logic error condition
 * - FATAL system error condition
 */
config.loglevel = 'NOTE'
```

## logListResult
```javascript
/**
 * If set to true, the results of each list call will be logged.
 * default: false
 * @type {boolean}
 */
config.logListResult = false
```

## truncateLogs
```javascript
/**
 * If set to true, string values will be truncated to `truncateStringLength` characters in the logs and array values will be trunctated to 30 elements.
 * default: false
 * @type {boolean}
 */
config.truncateLogs = false
```

## truncateStringLength
```javascript
/**
 * If `truncateLogs` is set to true, this is the maximum length of a string to be logged.
 * default: 100
 * @type {number}
 */
config.truncateStringLength = 100
```

## swaggerOptions
```javascript
/**
 * Set swagger options as per https://github.com/glennjones/hapi-swagger/blob/master/optionsreference.md
 * Options set here will override swagger config options below
 * @type {Object}
 config.swaggerOptions = {}
 ```

## docExpansion
```javascript
/**
 * Determines the initial expansion state of the swagger docs
 * - options: 'none', 'list', 'full' (default: 'none')
 * default: 'none'
 * @type {string}
 */
config.docExpansion = 'none'
```

## enableSwaggerUI
```javascript
/**
 * If set to false, SwaggerUI will not be generated.
 * - options: 'true', 'false' (default: 'true')
 * @type {boolean}
 */
config.enableSwaggerUI = true
```

## enableSwaggerHttps
```javascript
/**
 * If set to true, swagger will use the https protocol rather than http.
 * - options: 'true', 'false' (default: 'false')
 * @type {boolean}
 */
config.enableSwaggerHttps = false
```

## swaggerHost
```javascript
/**
 * Sets the host used for swagger requests. Useful for an api behind a reverse proxy.
 * default: undefined
 * @type {boolean}
 */
config.swaggerHost = undefined
```

 ## esm
```javascript
/**
 * If set to true, models will be loaded as esm modules (export default)
 * - options: 'true', 'false' (default: 'false')
 * @type {boolean}
 */
config.esm = false
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.2.0] - 2022-12-27
### Added
- [added post middleware for add and remove associations](https://github.com/JKHeadley/rest-hapi/pull/313)

## [3.1.0] - 2022-12-23
### Added
- [added option to truncate logs](https://github.com/JKHeadley/rest-hapi/pull/312)
### Fixed
- [remove dep on joi-objectid package](https://github.com/JKHeadley/rest-hapi/pull/303)

## [3.0.0] - 2022-08-23 [BREAKING]
### Changed
- [updated packages to remove vulnerabilities](https://github.com/JKHeadley/rest-hapi/pull/300)
- [BREAKING] Dropped support for node < 14

## [2.3.0] - 2021-05-03
### Added
- [added option to disable/enable list log](https://github.com/JKHeadley/rest-hapi/pull/270)

## [2.2.0] - 2020-07-08
### Added
- [allowList option to routeOptions](https://github.com/JKHeadley/rest-hapi/pull/244)

### Fixed
- [Mixed types to be any, not just object](https://github.com/JKHeadley/rest-hapi/pull/241)

## [2.1.0] - 2020-04-11
### Added
- [Mongoose config options](https://github.com/JKHeadley/rest-hapi/pull/239)

### Fixed
- [Added win support for loading Policies](https://github.com/JKHeadley/rest-hapi/pull/232)

## [2.0.0] - 2020-03-24 [BREAKING]
Please see the [release notes](https://github.com/JKHeadley/rest-hapi/issues/230).

## [1.9.0] - 2019-09-22 [BREAKING]
### Changed
- [BREAKING] [Updated $flatten query parameter to affect nested $embeds](https://github.com/JKHeadley/rest-hapi/pull/203)

## [1.8.0] - 2019-09-21
### Changed
- [Updated dependencies (hapi v18)](https://github.com/JKHeadley/rest-hapi/pull/201)

## [1.7.0] - 2019-09-13
### Added
- [Named parameters for wrapper methods](https://github.com/JKHeadley/rest-hapi/pull/200)
- [Simulated REST calls](https://github.com/JKHeadley/rest-hapi/pull/200)
- [Exposed `server` and `models` in RestHapi object](https://github.com/JKHeadley/rest-hapi/pull/200)

## [1.6.0] - 2019-05-06
### Added
- [Updated code for disabling authentication on specific endpoints.](https://github.com/JKHeadley/rest-hapi/pull/177)

### Fixed
- [Independent header validation for each endpoint](https://github.com/JKHeadley/rest-hapi/pull/176)

## [1.5.0] - 2019-04-30
### Added
- [Added general swagger options](https://github.com/JKHeadley/rest-hapi/pull/172)

### Changed
- [Updated dependencies](https://github.com/JKHeadley/rest-hapi/pull/146)

## [1.4.0] - 2019-03-18
### Added
- [feat: Add option to disable swagger](https://github.com/JKHeadley/rest-hapi/commit/b70a5360c40a6bce052cdc5937e24d9669b16e71)
- [update: allow 'id' user defined property in schema](https://github.com/JKHeadley/rest-hapi/pull/169)
- [Filter out fields postfixed with $\* ](https://github.com/JKHeadley/rest-hapi/pull/167)

### Changed
- [Bump minimatch to ^3.0.4 because of vulnerability](https://github.com/JKHeadley/rest-hapi/commit/8931d8d229dad80e08386ecc178c32ca8684128b)

### Fixed
- [fix undefined pointer when enabling auditLogScope](https://github.com/JKHeadley/rest-hapi/commit/d2e1340957d6620f3ab0d3b450a927e2859e0379)


## [1.3.0] - 2018-06-14
### Added
- Added 'swaggerHost' config option.

### Changed
- Removed bcrypt dependency

### Fixed
- Attempt to fix windows error

## [1.2.0] - 2018-05-31
### Changed
- Updated dependencies to fix security vulnerability.
- Refactored logging to be more consistent.

## [1.0.0] - 2018-05-27 [BREAKING]
Please see the [release notes](https://github.com/JKHeadley/rest-hapi/issues/120).

## [0.43.0] - 2018-03-23
### Changed
- Updated auditLog to get correct IP when behind a reverse proxy such as nginx.

## [0.42.0] - 2018-03-22
### Added
- Added option to toggle swagger protocol (http or https) with config.enableSwaggerHttps.

## [0.41.0] - 2018-03-17
### Added
- config.enableSwaggerUI to config.js to determine wether SwaggerUI will be generated or no.

## [0.40.0] - 2018-01-11
### Added
- [Config options](https://github.com/JKHeadley/rest-hapi/blob/master/README.md#configuration) for disabling `$where` query param and enabling mongoose [runValidators](http://mongoosejs.com/docs/validation.html#update-validators) option.

## [0.39.0] - 2017-12-22 [BREAKING]
### Added
- Added "remove: pre" middleware.

### Changed
- [BREAKING] Swapped parameters for "add: pre" middleware for consistency.

Before | After
--- | --- 
**add**    
pre(request, payload, Log) |      pre(payload, request, Log)  

## [0.38.0] - 2017-12-12
### Added
- [errorHelper](https://github.com/JKHeadley/rest-hapi#error-handling) feature.

### Changed
- Added ipAddress and notes to auditLog model
- Fixed date to set to actual current date in auditLog model
- Removed required fields in auditLog model
- updated errorHandler.formatResponse to handle boom errors and normal errors

## [0.37.0] - 2017-11-17
### Added
- [Duplicate fields](https://github.com/JKHeadley/rest-hapi#duplicate-fields) feature.
- Add [regex validator](https://github.com/JKHeadley/rest-hapi#route-validation).

### Fixed
- Moved testHelper to utilities directory to be included in npm module.
- Moved query-string to regular dependencies for testHelper.
- Updated travis since it broke for no apparent reason.
- Fixed bug preventing associations added from wrapper methods to update correctly.

## [0.36.0] - 2017-11-11
### Added
- Exposed testHelper to RestHapi object, mostly for server injection utilities.

### Fixed
- Fixed bug preventing \_id filtering for getAll queries with real ObjectIds.

## [0.35.0] - 2017-11-10
### Added
- Updated joi model for object types.

### Fixed
- Fixed bug preventing object types.

## [0.34.0] - 2017-11-05
### Added
- Added error response for db conflicts.

### Fixed
- Auditlog validation warning.
- Fix payload copies to work with ObjectIds.
- Added missing permissions.
- Added missing params to mongoose wrappers.

## [0.33.0] - 2017-10-23

config.js
/**
 * config.js - Configuration settings for the generated API
 */
const config = {}
config.mongo = {}

/**
 * Your app title goes here.
 * @type {string}
 */
config.appTitle = 'rest-hapi API'

/**
 * Your app version goes here.
 * @type {string}
 */
config.version = '1.0.0'

/**
 * Flag signifying whether the absolute path to the models directory is provided
 * default: false
 * @type {boolean}
 */
config.absoluteModelPath = false

/**
 * Path to the models directory
 * default: 'models'
 * @type {string}
 */
config.modelPath = 'models'

/**
 * Flag signifying whether the absolute path to the api directory is provided
 * @type {boolean}
 */
config.absoluteApiPath = false

/**
 * Path to the directory for additional endpoints
 * default: 'api'
 * @type {string}
 */
config.apiPath = 'api'

/**
 * Cors settings for generated endpoints. Can be set to false to disable.
 * @type {{additionalHeaders: string[], additionalExposedHeaders: string[]}}
 */
config.cors = {
  additionalHeaders: [],
  additionalExposedHeaders: []
}

/**
 * Mongo settings
 * - config.mongo.URI = 'mongodb://localhost/rest_hapi'; (local db, default)
 * - config.mongo.options = {} (mongoose options)
 */
config.mongo.URI = 'mongodb://localhost/rest_hapi'
config.mongo.options = {}

/**
 * Authentication strategy to be used for all generated endpoints.
 * Set to false for no authentication.
 * default: false
 * @type {boolean/string}
 */
config.authStrategy = false

/**
 * If set to false, MANY_MANY associations (including linking model data) will be saved in their own collection in th db.  This is useful if a single document
 * will be associated with many other documents, which could cause the document size to become very large. For example,
 * a business might be associated with thousands of users.
 *
 * Embedding the associations will be more efficient for population/association queries but less efficient for memory/document size.
 *
 * This setting can be individually overwritten by setting the "embedAssociation" association property.
 * default: false
 * @type {boolean}
 */
config.embedAssociations = false

/**
 * MetaData options:
 * - createdAt: (default: true) date specifying when the document was created.
 * - updatedAt: (default: true) date specifying when the document was last updated.
 * - deletedAt: (default: true) date specifying when the document was soft deleted.
 * - createdBy: (default: false) _id of user that created the document.
 * - updatedBy: (default: false) _id of user that last updated the document.
 * - updatedBy: (default: false) _id of user that soft deleted the document.
 */
config.enableCreatedAt = true
config.enableUpdatedAt = true
config.enableDeletedAt = true
config.enableCreatedBy = false
config.enableUpdatedBy = false
config.enableDeletedBy = false

/**
 * Enables raw $where mongoose queries
 * !!WARNING!!: This feature is meant for development ONLY and NOT in production as it provides direct query
 * access to your database.
 * default: false
 * @type {boolean}
 */
config.enableWhereQueries = false

/**
 * Enables fields from an associated model to be duplicated. Similar to permanently embedding an associated field within
 * the parent model schema. Useful if a parent model needs to be searchable or sortable by an association's field.
 * default: false
 * @type {boolean}
 */
config.enableDuplicateFields = false

/**
 * When true, duplicated fields will update whenever the original field is updated.
 * WARNING: This feature can make updates very resource intensive if many documents are duplicating the original field.
 * default: false
 * @type {boolean}
 */
config.trackDuplicatedFields = false

/**
 * When enabled, all create, update, associate, and delete events are recorded in an auditLog collection.
 * default: true
 * @type {boolean}
 */
config.enableAuditLog = true

/**
 * Values added here will be applied to the scope of the auditLog endpoint.
 * default: []
 * @type {Array}
 */
config.auditLogScope = []

/**
 * Specifies the TTL (time to live/lifetime/expiration) of auditLog documents. Accepts values in seconds unless specified
 * (Ex: 60 = 60 seconds, '1m' = 1 minute, or '1d' = 1 day)
 * See: http://nicoll.io/mongottl/
 * default: null (does not expire)
 * @type {string}
 */
config.auditLogTTL = null

/**
 * Enables policies via mrhorse (https://github.com/mark-bradshaw/mrhorse).
 * default: false
 * @type {boolean}
 */
config.enablePolicies = false

/**
 * Flag signifying whether the absolute path to the policies directory is provided.
 * default: false
 * @type {boolean}
 */
config.absolutePolicyPath = false

/**
 * Path to the directory for mrhorse policies (https://github.com/mark-bradshaw/mrhorse).
 * default: 'policies'
 * @type {string}
 */
config.policyPath = 'policies'

/**
 * Enables document level authorization.
 * default: true
 * @type {boolean}
 */
config.enableDocumentScopes = true

/**
 * If true, modifies the root scope of any document to allow access to the document's creator.
 * The scope value added is in the form: "user-{_id}" where "{_id}" is the _id of the user.
 * NOTE:
 * - This assumes that your authentication credentials (request.auth.credentials) will contain either
 * a "user" object with a "_id" property, or the user's _id stored in a property defined by "config.userIdKey".
 * - This also assumes that the user creating the document will have "user-{_id}" within their scope.
 * - Requires "config.enableDocumentScopes" to be "true".
 * - This setting can be individually overwritten by setting the "authorizeDocumentCreator" routeOptions property.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreator = false

/**
 * Same as "authorizeDocumentCreator", but modifies the "readScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToRead = false

/**
 * Same as "authorizeDocumentCreator", but modifies the "updateScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToUpdate = false

/**
 * Same as "authorizeDocumentCreator", but modifies the "deleteScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToDelete = false

/**
 * Same as "authorizeDocumentCreator", but modifies the "associateScope" rather than the root scope.
 * default: false
 * @type {boolean}
 */
config.authorizeDocumentCreatorToAssociate = false

/**
 * This is the path/key to the user _id stored in your request.auth.credentials object.
 * default: "user._id"
 * @type {string}
 */
config.userIdKey = 'user._id'

/**
 * Determines what action takes place when one or more document scope checks fail for requests dealing with multiple
 * documents (Ex: deleteMany or list). Options are:
 * - true: if one or more documents fail, the request responds with a 403.
 * - false: documents that don't pass are simply removed from the request (Ex: not deleted or not retrieved)
 * default: false
 * @type {boolean}
 */
config.enableDocumentScopeFail = false

/**
 * Flag specifying whether to text index all string fields for all models to enable text search.
 * WARNING: enabling this adds overhead to add inserts and updates, as well as added storage requirements.
 * default: false.
 * @type {boolean}
 */
config.enableTextSearch = false

/**
 * Soft delete options
 * - enableSoftDelete: adds "isDeleted" property to each model. Delete endpoints set "isDeleted" to true
 * unless the payload contains { hardDelete: true }, in which case the document is actually deleted (default false)
 * - filterDeletedEmbeds: if enabled, associations with "isDeleted" set to true will not populate (default false)
 * NOTE: this option is known to be buggy
 * @type {boolean}
 */
config.enableSoftDelete = false
config.filterDeletedEmbeds = false

/**
 * Validation options:
 * default: true
 * @type {boolean}
 */
config.enableQueryValidation = true
config.enablePayloadValidation = true
config.enableResponseValidation = true

/**
 * Mongoose validation options:
 * - enableMongooseRunValidators: enables the runValidators option in Mongoose update calls
 * <http://mongoosejs.com/docs/validation.html#update-validators>
 * default: false
 * @type {boolean}
 */
config.enableMongooseRunValidators = false

/**
 * Determines the hapi failAction of each response. Options are:
 * - true: responses that fail validation will return a 500 error.
 * - false: responses that fail validation will just log the offense and send the response as-is.
 * default: false
 * @type {boolean}
 */
config.enableResponseFail = false

/**
 * If set to true, (and authStrategy is not false) then endpoints will be generated with pre-defined
 * scopes based on the model definition.
 * default: false
 * @type {boolean}
 */
config.generateRouteScopes = false

/**
 * If set to true, the scope for each endpoint will be logged when then endpoint is generated.
 * default: false
 * @type {boolean}
 */
config.logScopes = false

/**
 * If set to true, each route will be logged as it is generated.
 * default: false
 * @type {boolean}
 */
config.logRoutes = false

/**
 * Log level options:
 * - INTERNAL use it for logging calls and other internal stuff
 * - DEBUG recommended to use it for debugging applications
 * - NOTE development verbose information (default)
 * - INFO minor information
 * - LOG significant messages
 * - WARNING really important stuff
 * - ERROR application business logic error condition
 * - FATAL system error condition
 */
config.loglevel = 'NOTE'

/**
 * If set to true, the results of each list call will be logged.
 * default: false
 * @type {boolean}
 */
config.logListResult = false

/**
 * If set to true, string values will be truncated to `truncateStringLength` characters in the logs and array values will be trunctated to 30 elements.
 * default: false
 * @type {boolean}
 */
config.truncateLogs = false

/**
 * If `truncateLogs` is set to true, this is the maximum length of a string to be logged.
 * default: 100
 * @type {number}
 */
config.truncateStringLength = 100

/**
 * Set swagger options as per https://github.com/glennjones/hapi-swagger/blob/master/optionsreference.md
 * Options set here will override swagger config options below
 *
 * @type {Object}
 */
config.swaggerOptions = {}

/**
 * Determines the initial expansion state of the swagger docs
 * - options: 'none', 'list', 'full' (default: 'none')
 * default: 'none'
 * @type {string}
 */
config.docExpansion = 'none'

/**
 * If set to false, SwaggerUI will not be generated.
 * - options: 'true', 'false' (default: 'true')
 * @type {boolean}
 */
config.enableSwaggerUI = true

/**
 * If set to true, swagger will use the https protocol rather than http.
 * - options: 'true', 'false' (default: 'false')
 * @type {boolean}
 */
config.enableSwaggerHttps = false

/**
 * Sets the host used for swagger requests. Useful for an api behind a reverse proxy.
 * default: undefined
 * @type {boolean}
 */
config.swaggerHost = undefined

/**
 * If set to true, models will be loaded as esm modules (export default)
 * - options: 'true', 'false' (default: 'false')
 * @type {boolean}
 */
config.esm = false

module.exports = config
